As I said, I think fc7ce2f is the wrong workaround. We should add a workaround to the TreePathPrefixMapper so it works in Optional rather than changing the APIs.

Error was:
```
19:56:20 In file included from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/STLExtras.h:19:0,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:15,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/TableGen/ScanDependencies.h:12,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:9:
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h: In instantiation of 'void llvm::optional_detail::OptionalStorage<T, <anonymous> >::emplace(Args&& ...) [with Args = {const llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>&}; T = llvm::BumpPtrAllocatorImpl<>; bool <anonymous> = false]':
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:73:14:   required from 'constexpr llvm::optional_detail::OptionalStorage<T, <anonymous> >::OptionalStorage(const llvm::optional_detail::OptionalStorage<T, <anonymous> >&) [with T = llvm::BumpPtrAllocatorImpl<>; bool <anonymous> = false]'
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:248:13:   required from 'struct std::is_trivially_move_constructible<llvm::TreePathPrefixMapper>'
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:55:79:   required from 'class llvm::Optional<llvm::TreePathPrefixMapper>'
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:219:34:   required from here
19:56:20 
/home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/ADT/Optional.h:112:5: error: use of deleted function 'llvm::BumpPtrAllocatorImpl<>::BumpPtrAllocatorImpl(const llvm::BumpPtrAllocatorImpl<>&)'
19:56:20      ::new ((void *)std::addressof(value)) T(std::forward<Args>(args)...);
19:56:20      ^
19:56:20 In file included from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/ThreadSafeAllocator.h:12:0,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/FileSystemCache.h:18,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/CAS/CachingOnDiskFileSystem.h:12,
19:56:20                  from /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/lib/TableGen/ScanDependencies.cpp:10:
19:56:20 /home/buildnode/jenkins/workspace/apple-llvm-project-pr-linux/llvm-project/llvm/include/llvm/Support/Allocator.h:67:7: note: 'llvm::BumpPtrAllocatorImpl<>::BumpPtrAllocatorImpl(const llvm::BumpPtrAllocatorImpl<>&)' is implicitly declared as deleted because 'llvm::BumpPtrAllocatorImpl<>' declares a move constructor or move assignment operator
19:56:20  class BumpPtrAllocatorImpl
19:56:20        ^
```
Looks like the probably is that `BumpPtrAllocatorImpl`'s copy constructor is being called for some reason. May need to adjust the TreePathPrefixMapper constructor to be more explicit in some way.

--- a/clang/include/clang/Tooling/DependencyScanning/DependencyScanningFilesystem.h
+++ b/clang/include/clang/Tooling/DependencyScanning/DependencyScanningFilesystem.h
@@ -84,17 +84,17 @@ private:
 
   IntrusiveRefCntPtr<llvm::cas::CASFileSystemBase> FS;
 
-  struct Entry {
+  struct FileEntry {
     std::error_code EC; // If non-zero, caches a stat failure.
     Optional<StringRef> Buffer;
     llvm::vfs::Status Status;
     Optional<llvm::cas::CASID> ID;
   };
   llvm::BumpPtrAllocator EntryAlloc;
-  llvm::StringMap<Entry, llvm::BumpPtrAllocator> Entries;
+  llvm::StringMap<FileEntry, llvm::BumpPtrAllocator> Entries;
 
   struct LookupPathResult {
-    const Entry *Entry = nullptr;
+    const FileEntry *Entry = nullptr;
 
     // Only filled if the Entry is nullptr.
     llvm::ErrorOr<llvm::vfs::Status> Status;

--- a/llvm/include/llvm/ADT/Optional.h
+++ b/llvm/include/llvm/ADT/Optional.h
@@ -34,12 +34,12 @@ namespace optional_detail {
 /// Storage for any type.
 //
 // The specialization condition intentionally uses
-// llvm::is_trivially_copy_constructible instead of
-// std::is_trivially_copy_constructible.  GCC versions prior to 7.4 may
-// instantiate the copy constructor of `T` when
-// std::is_trivially_copy_constructible is instantiated.  This causes
-// compilation to fail if we query the trivially copy constructible property of
-// a class which is not copy constructible.
+// llvm::is_trivially_{copy/move}_constructible instead of
+// std::is_trivially_{copy/move}_constructible. GCC versions prior to 7.4 may
+// instantiate the copy/move constructor of `T` when
+// std::is_trivially_{copy/move}_constructible is instantiated.  This causes
+// compilation to fail if we query the trivially copy/move constructible
+// property of a class which is not copy/move constructible.
 //
 // The current implementation of OptionalStorage insists that in order to use
 // the trivial specialization, the value_type must be trivially copy
@@ -50,12 +50,13 @@ namespace optional_detail {
 //
 // The move constructible / assignable conditions emulate the remaining behavior
 // of std::is_trivially_copyable.
-template <typename T, bool = (llvm::is_trivially_copy_constructible<T>::value &&
-                              std::is_trivially_copy_assignable<T>::value &&
-                              (std::is_trivially_move_constructible<T>::value ||
-                               !std::is_move_constructible<T>::value) &&
-                              (std::is_trivially_move_assignable<T>::value ||
-                               !std::is_move_assignable<T>::value))>
+template <typename T,
+          bool = (llvm::is_trivially_copy_constructible<T>::value &&
+                  std::is_trivially_copy_assignable<T>::value &&
+                  (llvm::is_trivially_move_constructible<T>::value ||
+                   !std::is_move_constructible<T>::value) &&
+                  (std::is_trivially_move_assignable<T>::value ||
+                   !std::is_move_assignable<T>::value))>
 class OptionalStorage {
   union {
     char empty;

--- a/llvm/include/llvm/CAS/OnDiskHashMappedTrie.h
+++ b/llvm/include/llvm/CAS/OnDiskHashMappedTrie.h
@@ -57,7 +57,7 @@ public:
 
     explicit operator bool() const { return I == -2u; }
 
-    LookupResult() noexcept = default;
+    LookupResult() = default;
     LookupResult(LookupResult &&) = default;
     LookupResult &operator=(LookupResult &&) = default;
 

--- a/llvm/lib/CAS/CASFileSystem.cpp
+++ b/llvm/lib/CAS/CASFileSystem.cpp
@@ -348,7 +348,7 @@ static Expected<std::unique_ptr<CASFileSystem>>
 initializeCASFileSystem(std::unique_ptr<CASFileSystem> FS, CASID RootID) {
   if (Error E = FS->initialize(RootID))
     return std::move(E);
-  return FS;
+  return std::move(FS);
 }
 
 Expected<std::unique_ptr<CASFileSystemBase>>

--- a/llvm/lib/CAS/HashMappedTrie.cpp
+++ b/llvm/lib/CAS/HashMappedTrie.cpp
@@ -63,9 +63,11 @@ private:
 };
 } // namespace
 
+namespace llvm {
 template <> struct isa_impl<ThreadSafeSubtrie, TrieNode> {
   static inline bool doit(const TrieNode &TN) { return TN.IsSubtrie; }
 };
+} // namespace llvm
 
 std::unique_ptr<ThreadSafeSubtrie> ThreadSafeSubtrie::create(size_t StartBit,
                                                              size_t NumBits) {

--- a/llvm/lib/CASObjectFormats/FlatV1.cpp
+++ b/llvm/lib/CASObjectFormats/FlatV1.cpp
@@ -25,12 +25,12 @@ using namespace llvm;
 using namespace llvm::casobjectformats;
 using namespace llvm::casobjectformats::flatv1;
 
-const StringLiteral CompileUnitRef::KindString;
-const StringLiteral NameRef::KindString;
-const StringLiteral SectionRef::KindString;
-const StringLiteral BlockRef::KindString;
-const StringLiteral SymbolRef::KindString;
-const StringLiteral BlockContentRef::KindString;
+constexpr StringLiteral CompileUnitRef::KindString;
+constexpr StringLiteral NameRef::KindString;
+constexpr StringLiteral SectionRef::KindString;
+constexpr StringLiteral BlockRef::KindString;
+constexpr StringLiteral SymbolRef::KindString;
+constexpr StringLiteral BlockContentRef::KindString;
 
 void ObjectFileSchema::anchor() {}
 

--- a/llvm/lib/CASObjectFormats/NestedV1.cpp
+++ b/llvm/lib/CASObjectFormats/NestedV1.cpp
@@ -21,15 +21,15 @@ using namespace llvm;
 using namespace llvm::casobjectformats;
 using namespace llvm::casobjectformats::nestedv1;
 
-const StringLiteral NameRef::KindString;
-const StringLiteral BlockDataRef::KindString;
-const StringLiteral TargetListRef::KindString;
-const StringLiteral SectionRef::KindString;
-const StringLiteral BlockRef::KindString;
-const StringLiteral SymbolRef::KindString;
-const StringLiteral SymbolTableRef::KindString;
-const StringLiteral NameListRef::KindString;
-const StringLiteral CompileUnitRef::KindString;
+constexpr StringLiteral NameRef::KindString;
+constexpr StringLiteral BlockDataRef::KindString;
+constexpr StringLiteral TargetListRef::KindString;
+constexpr StringLiteral SectionRef::KindString;
+constexpr StringLiteral BlockRef::KindString;
+constexpr StringLiteral SymbolRef::KindString;
+constexpr StringLiteral SymbolTableRef::KindString;
+constexpr StringLiteral NameListRef::KindString;
+constexpr StringLiteral CompileUnitRef::KindString;
 
 namespace {
 class EncodedDataRef : public SpecificRef<EncodedDataRef> {
@@ -55,7 +55,7 @@ private:
 };
 } // end namespace
 
-const StringLiteral EncodedDataRef::KindString;
+constexpr StringLiteral EncodedDataRef::KindString;
 
 void ObjectFileSchema::anchor() {}
 

--- a/llvm/tools/llvm-libtool-darwin/llvm-libtool-darwin.cpp
+++ b/llvm/tools/llvm-libtool-darwin/llvm-libtool-darwin.cpp
@@ -632,7 +632,7 @@ static Expected<Config> parseCommandLine(int Argc, char **Argv) {
       LibraryOperation.error("must be specified", "", Stream);
       return createStringError(std::errc::invalid_argument, Error.c_str());
     }
-    return C;
+    return std::move(C);
   }
 
   if (OutputFile.empty()) {
@@ -683,7 +683,7 @@ static Expected<Config> parseCommandLine(int Argc, char **Argv) {
     C.CAS = std::move(*MaybeCAS);
   }
 
-  return C;
+  return std::move(C);
 }
 
 int main(int Argc, char **Argv) {

--- a/llvm/unittests/ADT/OptionalTest.cpp
+++ b/llvm/unittests/ADT/OptionalTest.cpp
@@ -805,4 +805,15 @@ TEST(OptionalTest, HashValue) {
   EXPECT_EQ(hash_value(B), hash_value(I));
 }
 
+struct NotTriviallyCopyable {
+  NotTriviallyCopyable(); // Constructor out-of-line.
+  virtual ~NotTriviallyCopyable() = default;
+  Optional<MoveOnly> MO;
+};
+
+TEST(OptionalTest, GCCIsTriviallyMoveConstructibleCompat) {
+  Optional<NotTriviallyCopyable> V;
+  EXPECT_FALSE(V);
+}
+
 } // end anonymous namespace

